### PR TITLE
ICU-22787 Fix ClangCL compilation on Windows

### PR DESCRIPTION
**Pull request by StefanStojanovic**
I'm working on supporting Node.js on Windows. There is an ongoing effort to enable it to be compiled with ClangCL. While working on it I noticed an issue when creating `.obj` file in ICU because of the `*pCPU = IMAGE_FILE_MACHINE_UNKNOWN;` line, and the code I made fixes that by adding a new option. Although this is a broader problem that should be fixed for multiple compilers and platforms, my change is aimed toward ClangCL and Windows, but can be used as a good foundation to work on fixes for other platforms and compilers and make corrections in future iterations.

This issue is currently our main blocker for enabling ClangCL compilation for Node.js and it'd be greatly appreciated if it could be reviewed soon.

Refs: https://github.com/nodejs/node/pull/53003

##### Checklist

- [x] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-22787
- [x] Required: The PR title must be prefixed with a JIRA Issue number. 
- [x] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [x] Required: Each commit message must be prefixed with a JIRA Issue number. 
- [x] Issue accepted (done by Technical Committee after discussion)
- [ ] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable

### CODE REVIEW

**Comment by srl295** [comment]
#3013

File: icu4c/source/tools/toolutil/pkg_genc.cpp
Comment (roubert) [inline icu4c/source/tools/toolutil/pkg_genc.cpp:867]: In that case, I'd say that it'd be easier to debug if the process terminated with an error message if that which should never happen ever happened anyway.

File: icu4c/source/tools/genccode/genccode.c
Comment (roubert) [inline icu4c/source/tools/genccode/genccode.c:136]: Wouldn't it be easier to use this tool correctly if this commandline flag was `#ifdef`'ed out altogether on platforms where it isn't implemented?

File: icu4c/source/tools/toolutil/pkg_genc.cpp
Comment (roubert) [inline icu4c/source/tools/toolutil/pkg_genc.cpp:855]: If I read the existing `U_PLATFORM_HAS_WIN32_API` block correctly it's all about Microsoft's `link.exe`, so it seems odd to now mix in another linker in that. Wouldn't this become easier to read if Microsoft's and LLVM's linkers got each their separate `#if` block?

File: icu4c/source/tools/toolutil/pkg_genc.cpp
Comment (roubert) [inline icu4c/source/tools/toolutil/pkg_genc.cpp:863]: Are 32-bit builds not supported? It's been a long time since I last had any reason to do one, but I wasn't aware that they're no longer supported.

**Review by roubert — CHANGES_REQUESTED** [review]

File: icu4c/source/tools/toolutil/pkg_genc.cpp
Comment (StefanStojanovic) [reply icu4c/source/tools/toolutil/pkg_genc.cpp:869]: Makes sense. I've changed it.

File: icu4c/source/tools/genccode/genccode.c
Comment (StefanStojanovic) [reply icu4c/source/tools/genccode/genccode.c:141]: For now, I've only `#ifdef`ed out the parts in `genccode.c`, as I didn't want to make `writeObjectCode` signature different depending on which compiler is used on Windows. I just pass it `NULL` if Clang isn't used. If you think I should make the signature `#ifdef`ed too, let me know.

File: icu4c/source/tools/toolutil/pkg_genc.cpp
Comment (StefanStojanovic) [reply icu4c/source/tools/toolutil/pkg_genc.cpp:855]: In `platform.h` for `U_PLATFORM_HAS_WIN32_API` it says: `Defines whether the Win32 API is available on the platform.`. This is true for ClangCL, so I've added a new constant `U_CLANG_CL` to distinguish when using Clang on Windows. After that, I changed to using that constant inside the mentioned block.

> Wouldn't this become easier to read if Microsoft's and LLVM's linkers got each their separate #if block

It would, but I do not know how to write linker-specific code (I'm not saying it is impossible). What can be done from my perspective is to detect the compiler (MSVC or Clang) and then infer a linker (link.exe or lld-link.exe), which is more or less what we have now after these changes.

File: icu4c/source/tools/toolutil/pkg_genc.cpp
Comment (StefanStojanovic) [reply icu4c/source/tools/toolutil/pkg_genc.cpp:863]: Makes sense. I've added a `strcmp(optCpuArch, "x86") == 0` branch in case someone needs it since it is still supported.

Above this code there is:

```
#   if defined(_M_IX86)
        *pCPU = IMAGE_FILE_MACHINE_I386;
```

but that would only apply if `genccode` was compiled targeting x86, which is not necessarily the case.

**Comment by StefanStojanovic** [comment]
Thanks for the review @roubert, I tried to address all your comments in a new commit.

I see 2 checks failed: `jira-ticket` and `single-commit`. I'll squash commits and keep only the original commit message once all is ready to land, but I kept 2 commits now for easier reviewing.

**Comment by markusicu** [comment]
> I see 2 checks failed: `jira-ticket` and `single-commit`. I'll squash commits and keep only the original commit message once all is ready to land, but I kept 2 commits now for easier reviewing.

Please wait with squashing until after feedback is resolved.

**Comment by StefanStojanovic** [comment]
> Please wait with squashing until after feedback is resolved.

Will do.

By the way, I see that in `CI-ICU4C` all `MSVC` builds failed. I'm not sure why that happened, but I built it locally before pushing (`allinone.sln` -> Build Solution) and it worked.

File: icu4c/source/tools/genccode/genccode.c
Comment (srl295) [reply icu4c/source/tools/genccode/genccode.c:141]: I think it would be better to leave it a generic argument, even if it is only currently valid for ClangCL.  It may be needed in other situations.

**Comment by StefanStojanovic** [comment]
Hey everyone, I wanted to check if there is anything else for me to do after my initial changes? If you think it is better to leave the new option I can every my latest changes in `genccode.c`. Other than that, are the other changes good? Regards.

**Comment by roubert** [comment]
> I see that in `CI-ICU4C` all `MSVC` builds failed. I'm not sure why that happened,

The CI is currently broken, you'll just have to wait for it to get fixed (and then those tests should pass again).

**Comment by roubert** [comment]
> The CI is currently broken,

It has now been fixed and the tests pass.

**Review by roubert — CHANGES_REQUESTED** [review]

File: icu4c/source/common/unicode/platform.h
Comment (StefanStojanovic) [reply icu4c/source/common/unicode/platform.h:227]: I agree, I'll remove it.

File: icu4c/source/tools/toolutil/pkg_genc.cpp
Comment (StefanStojanovic) [reply icu4c/source/tools/toolutil/pkg_genc.cpp:869]: I added it based on your previous comment:

> In that case, I'd say that it'd be easier to debug if the process terminated with an error message if that which should never happen ever happened anyway.

Although I didn't think about it before you mentioned it, this log makes sense to me. For example, imagine someone makes a typo and passes `-c X64` or `-c amr64`, this log would help them fix it much quicker.

Comment (roubert) [reply icu4c/source/tools/toolutil/pkg_genc.cpp:869]: > For example, imagine someone makes a typo and passes `-c X64` or `-c amr64`, this log would help them fix it much quicker.

Wait a second, your original comment here was _"This should never happen."_ but the user making a typo on the commandline is the total opposite of that, that's something that certainly is going to happen, sooner or later.

But verifying that the commandline arguments are correct should be done already in `main()` in `genccode.c`, in the same way that the arguments to `-a` are being verified there by calling the `checkAssemblyHeaderName()` helper function.

**Review by roubert — CHANGES_REQUESTED** [review]

File: icu4c/source/tools/toolutil/pkg_genc.cpp
Comment (StefanStojanovic) [reply icu4c/source/tools/toolutil/pkg_genc.cpp:869]: I understand the confusion. My comment was misleading. Sorry for that. I checked the code now and saw the `checkAssemblyHeaderName` helper, I'll add that type of check to the `-c` option as well. Thanks for pointing that out. Once I add that the else branch should really _**never happen**_, and then the change you suggested with `std::terminate()` will be correct.

**Comment by StefanStojanovic** [comment]
@roubert I pushed the changes in a new commit. I addressed your comments and made changes based on [this comment](https://github.com/unicode-org/icu/pull/3023#discussion_r1631634392). Please let me know if there is something else I should change. Regards.

**Review by roubert — COMMENTED** [review]
OK, this all looks reasonable to me now, so I'll leave it to @rp9-next (@microsoft) to take this review from here (and verify that what this does really is what it should be doing, and that this is the preferred way of doing that, and so on).

**Review by rp9-next — APPROVED** [review]

**Comment by rp9-next** [comment]
@roubert , merging is blocked by a change requested check which I believe is outdated.. 
Could you dismiss it ?

@StefanStojanovic, please squash all the commits to a single commit for merging (you could use the[ squash branch tool](https://us-central1-dev-infra-273822.cloudfunctions.net/unicode-github-bot/info/unicode-org/icu/3023) as well)

**Pull request merged**
